pingcap/tidb-operator
# supoport canary deployment

### What problem does this PR solve? 
https://github.com/pingcap/tidb-operator/pull/3243
https://github.com/pingcap/tidb-operator/pull/3166
https://github.com/pingcap/tidb-operator/pull/1866

### What is changed and how does it work?
This PR is merged with https://github.com/pingcap/tidb-operator/pull/3548 ,  @shonge and I did this together. **Please review the code of `helm` part in his PR https://github.com/pingcap/tidb-operator/pull/3548.**
Users now can deploy mutiple operators(tidb-controller-manager and tidb-scheduler) in k8s by helm. And tidb-controller-manager will work independently by filtering the CR by label.

### Check List 

Tests 

- Manual test 
  - Install two operator releases by helm with values of `scheduler.serviceAccount`, `scheduler.schedulerName`, `controllerManager.serviceAccount` different and only install `controllerManager` and `scheduler`. (`advancedStatefulset` and `admissionWebhook` is not allowed to install mutiple) And one controller is set the `controllerManager.selector` as `k1=v1`, another is set as `k1!=v1`.
  - Create two `TidbCluster`, one is labeled with `k1=v1` and set the scheduler name as `tidb-scheduler-1`, another is not labeled and set the scheduler name as `tidb-scheduler-2`.
  - Notice that each `TidbCluster` will **only** be watched by specific controller.


Code changes

- Has Go code change

Related changes

- Need to update the documentation

### Does this PR introduce a user-facing change?:

```release-note
Support configuring `--selector` for the tidb-controller-manager to manage specific CRs 
```
